tls/src/server.rs

```diff
@@ -0,0 +1,142 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! A TLS based server
+
+use crate::load_root_cert;
+use camino::Utf8PathBuf;
+use rustls::version::TLS13;
+use rustls::{
+    server::{
+        danger::{ClientCertVerified, ClientCertVerifier},
+        ResolvesServerCert,
+    },
+    CipherSuite, ServerConfig, SignatureScheme,
+};
+use slog::error;
+use std::sync::Arc;
+
+use crate::{crypto_provider, LocalCertResolver, RotCertVerifier};
+
+impl ResolvesServerCert for LocalCertResolver {
+    fn resolve(
+        &self,
+        client_hello: rustls::server::ClientHello,
+    ) -> Option<Arc<rustls::sign::CertifiedKey>> {
+        // We only support Ed25519 and we control both sides of the connection
+        if !client_hello
+            .signature_schemes()
+            .iter()
+            .all(|&s| s == SignatureScheme::ED25519)
+        {
+            error!(
+                self.log,
+                "Invalid signature scheme(s) in client hello message: {:?}",
+                client_hello.signature_schemes()
+            );
+            return None;
+        }
+
+        // We only want to allow `TLS13_CHACHA20_POLY1305_SHA256`
+        // from the client, but rustls automatically inserts
+        // `TLS_EMPTY_RENEGOTIATION_INFO_SCSV` into the `ClientHello`. Therefore
+        // we filter before checking for our desired algorithm.
+        if !client_hello
+            .cipher_suites()
+            .iter()
+            .filter(|&&s| s != CipherSuite::TLS_EMPTY_RENEGOTIATION_INFO_SCSV)
+            .all(|&s| s == CipherSuite::TLS13_CHACHA20_POLY1305_SHA256)
+        {
+            error!(
+                self.log,
+                "Invalid cipher suite(s) in client hello message: {:?}",
+                client_hello.cipher_suites()
+            );
+            return None;
+        }
+
+        match self.load_certified_key() {
+            Ok(key) => Some(key),
+            Err(e) => {
+                error!(self.log, "failed to load certified key: {e}");
+                None
+            }
+        }
```

Comment:
nitpicky, take it or leave it: isn't this `Option::ok_or_else`?

Reply:
I actually don't think it makes sense to do that with side effects like the `error!` log. I feel like clippy has yelled at me for things like this in the past.